### PR TITLE
GatherPileupSummaries

### PR DESCRIPTION
**Pull request by takutosato**
@davidbenjamin will you review? Scatter-gather works as expected on the orientation bias evaluation samples.

### CODE REVIEW

**Comment by davidbenjamin** [comment]
As discussed offline, let's put pileup summaries inside the M2 task, but later leave CollectF1R2 in its own separate task.  Otherwise looks good.

**Comment by takutosato** [comment]
@davidbenjamin `LearnReadOrientationModel` now takes in lists of inputs, which allows for parallelization of `CollectF1R2Counts.` The updated wdl includes this change. I also moved `GetPileupSumamries` to the M2 task, as discussed offline. All the new changes are in the third commit, "update the wdl, prallelize..."

File: scripts/mutect2_wdl/mutect2.wdl [hunk 1]
```diff
@@ -194,6 +194,38 @@ workflow Mutect2 {
...
+                    intervals = subintervals,
+                    extra_intervals = ob_mm_filter_training_intervals,
+                    max_retries = max_retries
+            }
+        }
+
+        call LearnReadOrientationModel {
+            input:
+                alt_tables = CollectF1R2Counts.alt_table,
+                ref_histograms = CollectF1R2Counts.ref_histogram,
```
Comment (davidbenjamin) [inline scripts/mutect2_wdl/mutect2.wdl]: ref has plural "histograms"  on LHS and singular "histogram" on RHS; alt has plural on both sides.

File: scripts/mutect2_wdl/mutect2.wdl [hunk 2]
```diff
@@ -575,6 +606,21 @@ task M2 {
...
+        ### GetPileupSummaries
+        # These must be created, even if they remain empty, as cromwell doesn't support optional output 
+        touch tumor-pileups.table
+        touch normal-pileups.table
+
+        if [[ -f "${variants_for_contamination}" ]]; then
+            gatk --java-options "-Xmx${command_mem}m" GetPileupSummaries -R ${ref_fasta} -I ${tumor_bam} ${"--interval-set-rule INTERSECTION -L " + intervals} \
+                -V ${variants_for_contamination} -L ${variants_for_contamination} -O tumor-pileups.table
+
+            if [[ -f ${normal_bam} ]]; then
```
Comment (davidbenjamin) [inline scripts/mutect2_wdl/mutect2.wdl]: Does this logic work with NIO where the normal bam isn't localized?  I would think `[[ -f ]]` would give false.

Comment (davidbenjamin) [reply scripts/mutect2_wdl/mutect2.wdl]: If this is the case and you have a fix I would do the same for variants for contamination.  It's currently a `File` but we want to future-proof.

File: scripts/mutect2_wdl/mutect2.wdl [hunk 3]
```diff
@@ -700,6 +748,51 @@ task MergeBamOuts {
...
+    # Mem is in units of GB but our command and memory runtime values are in MB
+    Int machine_mem = if defined(mem) then mem * 1000 else 3500
+    Int command_mem = machine_mem - 1000
+
+    command {
+        set -e
+        export GATK_LOCAL_JAR=${default="/root/gatk.jar" gatk_override}
+
+        gatk --java-options "-Xmx${command_mem}m" GatherPileupSummaries \
+        --sequence-dictionary ${ref_dict} \
```
Comment (davidbenjamin) [inline scripts/mutect2_wdl/mutect2.wdl]: indent

File: scripts/mutect2_wdl/mutect2.wdl [hunk 4]
```diff
@@ -700,6 +748,51 @@ task MergeBamOuts {
...
+        -I ${sep=' -I ' input_tables} \
```
Comment (davidbenjamin) [inline scripts/mutect2_wdl/mutect2.wdl]: Double-checking: is this error-free when input tables is empty or not present?

File: scripts/mutect2_wdl/mutect2.wdl [hunk 5]
```diff
@@ -831,11 +925,15 @@ task LearnReadOrientationModel {
...
+        if [[ "${alt_hist_length}" > 0 ]]; then
+            ah_flag="-ah"
+        fi
+
         gatk --java-options "-Xmx${command_mem}m" LearnReadOrientationModel \
-        -alt-table ${alt_table} \
-        -ref-hist ${ref_histogram} \
-        -alt-hist ${alt_histograms} \
-        -O "${tumor_sample}-artifact-prior-table.tsv"
+        -at ${sep=" -at " alt_tables} \
```
Comment (davidbenjamin) [inline scripts/mutect2_wdl/mutect2.wdl]: indent

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java [hunk 6]
```diff
@@ -98,6 +101,37 @@ public static void writeToFile(final String sample, final List<PileupSummary> re
...
+                try ( PileupSummaryTableReader reader = new PileupSummaryTableReader(inputFile)){
+                    if (! headerWritten){
+                        sample = reader.getMetadata().get(TableUtils.SAMPLE_METADATA_TAG);
+                        writer.writeMetadata(TableUtils.SAMPLE_METADATA_TAG, sample);
+                        headerWritten = true;
+                    }
+
+                    final String thisSample = reader.getMetadata().get(TableUtils.SAMPLE_METADATA_TAG);
+                    if (! thisSample.equals(sample)){
+                        throw new UserException(String.format("Combining PileupSummaryTables from different samples is not supported. Got samples %s and %s",
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java]: Make it a `UserException.BadInput`

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java [hunk 7]
```diff
@@ -107,6 +141,31 @@ public static void writeToFile(final String sample, final List<PileupSummary> re
...
     }
 
+    public static class PileupSummaryComparator implements Comparator<PileupSummary> {
+        final SAMSequenceDictionary sequenceDictionary;
+        final List<String> contigsInOrder;
+
+        public PileupSummaryComparator(final SAMSequenceDictionary sequenceDictionary){
+            this.sequenceDictionary = sequenceDictionary;
+            contigsInOrder = sequenceDictionary.getSequences().stream()
+                    .map(ssr -> ssr.getSequenceName())
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java]: This looks like alphabetical order; is that right?  If so, would the sequence dictionary's own order make more sense?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java [hunk 8]
```diff
@@ -107,6 +141,31 @@ public static void writeToFile(final String sample, final List<PileupSummary> re
...
+        }
+
+        @Override
+        public int compare(PileupSummary ps1, PileupSummary ps2) {
+            // Use Contig Index in case the contig name is e.g. chr2
+            final int contigIndex1 = contigsInOrder.indexOf(ps1.getContig());
+            final int contigIndex2 = contigsInOrder.indexOf(ps2.getContig());
+
+            if (contigIndex1 != contigIndex2){
+                return contigIndex1 - contigIndex2;
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java]: `Integer.compare` is clearer than subtraction.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java [hunk 9]
```diff
@@ -107,6 +141,31 @@ public static void writeToFile(final String sample, final List<PileupSummary> re
...
+            } else {
+                return ps1.getStart() - ps2.getStart();
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java]: Likewise.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/CollectF1R2Counts.java [hunk 10]
```diff
@@ -44,13 +44,16 @@
 @CommandLineProgramProperties(
         summary = "Collect F1R2 read counts for the Mutect2 orientation bias mixture model filter",
         oneLineSummary = "Collect F1R2 read counts for the Mutect2 orientation bias mixture model filter",
-        programGroup = CoverageAnalysisProgramGroup.class
+        programGroup = ShortVariantDiscoveryProgramGroup.class
 )
 
 public class CollectF1R2Counts extends LocusWalker {
     public static final String ALT_DATA_TABLE_LONG_NAME = "alt-table";
+    public static final String ALT_DATA_TABLE_SHORT_NAME = "at";
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/CollectF1R2Counts.java]: I think it's better not to define short names for these.  The long names are short enough.  Note that this changes your wdls, which used the short names.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java [hunk 11]
```diff
@@ -72,25 +80,22 @@
 
     @Override
     protected void onStartup(){
-        final MetricsFile<?, Integer> referenceSiteMetrics = readMetricsFile(refHistogramTable);
-        refHistograms = referenceSiteMetrics.getAllHistograms();
+        refHistograms = sumHistogramsFromFiles(refHistogramFiles);
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java]: You should validate that the input lists all have the same length.  If there's some check to make sure the headers' orders correspond that would be good too.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java [hunk 12]
```diff
@@ -247,12 +252,51 @@ public static void mergeDesignMatrices(final List<AltSiteRecord> altDesignMatrix
...
+    public static List<Histogram<Integer>> sumHistogramsFromFiles(final List<File> files){
+        Utils.nonNull(files, "files may not be null");
+
+        final List<Histogram<Integer>> histogramList = readMetricsFile(files.get(0)).getAllHistograms();
+        for (int i = 1; i < files.size(); i++){
+            final List<Histogram<Integer>> ithHistograms = readMetricsFile(files.get(i)).getAllHistograms();
+            for (final Histogram<Integer> jthHistogram : ithHistograms){
+                final String refContext = jthHistogram.getValueLabel();
+                final Optional<Histogram<Integer>> hist = histogramList.stream().filter(h -> h.getValueLabel().equals(refContext)).findAny();
+                if (! hist.isPresent()){
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java]: `Utils.validate(hist.isPresent())`

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java [hunk 13]
```diff
@@ -247,12 +252,51 @@ public static void mergeDesignMatrices(final List<AltSiteRecord> altDesignMatrix
...
+            final Pair<String, List<AltSiteRecord>> ithSampleAndRecords = AltSiteRecord.readAltSiteRecords(tables.get(i), defaultInitialListSize);
+            if (! sample.equals(ithSampleAndRecords.getLeft())){
+                throw new UserException("altDataTables contains records for two samples: " + sample + " and " + ithSampleAndRecords.getLeft());
+            }
+
+            records.addAll(ithSampleAndRecords.getRight());
+        }
+
+        return new ImmutablePair<>(sample, records);
+
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java]: extra whitespace

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java (hunk 6 above)
Comment (takutosato) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java (hunk 7 above)
Comment (takutosato) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java]: I'm using the order in the sequence dictionary, unless I'm missing implicit sorting that takes places somewhere

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java (hunk 8 above)
Comment (takutosato) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java]: Nice catch. Done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java (hunk 9 above)
Comment (takutosato) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/PileupSummary.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/CollectF1R2Counts.java (hunk 10 above)
Comment (takutosato) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/CollectF1R2Counts.java]: True. Done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java (hunk 11 above)
Comment (takutosato) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java]: Validate input list size, done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java (hunk 13 above)
Comment (takutosato) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java]: done

File: scripts/mutect2_wdl/mutect2.wdl (hunk 1 above)
Comment (takutosato) [reply scripts/mutect2_wdl/mutect2.wdl]: fixed

File: scripts/mutect2_wdl/mutect2.wdl (hunk 2 above)
Comment (takutosato) [reply scripts/mutect2_wdl/mutect2.wdl]: changed to `! -z`. That should work, will test.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java (hunk 11 above)
Comment (takutosato) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java]: I added a check for headers of the histograms, and `TableReader` will ensure that that `altSiteRecords` are formatted properly.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java (hunk 12 above)
Comment (takutosato) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/LearnReadOrientationModel.java]: done

**Comment by takutosato** [comment]
@davidbenjamin at long last, back to you. I updated the nio wdl too, and it passes the Firecloud M2 wdl validation with the HCC sample, but not with the cram. But that's because that cram file is aligned to hg38, whereas the workspace uses hg19. I didn't touch anything related to the CramToBam task in the nio wdl so I think we're OK.

**Comment by davidbenjamin** [comment]
@takutosato I figured it was a rebase thing.  You can merge once Travis passes.

**Pull request merged**